cmd/gateway-sia-cachelayer.go

```diff
@@ -0,0 +1,685 @@
+		return err
+	}
+
+	// Start the cache management process
+	cache.CacheTicker = time.NewTicker(time.Second * time.Duration(cache.ManagerDelaySec))
+	go func() {
+		for _ = range cache.CacheTicker.C {
+			cache.manager()
+		}
+	}()
+
+	return nil
+}
+
+// Stop will stop the SiaCacheLayer
+func (cache *SiaCacheLayer) Stop() {
+	cache.debugmsg("SiaCacheLayer.Stop")
+
+	// Stop cache management process
+	cache.CacheTicker.Stop()
+
+	// Close the database
+	cache.dbCloseDatabase()
+}
+
+// InsertBucket will attempt to insert a new bucket
+func (cache *SiaCacheLayer) InsertBucket(bucket string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.InsertBucket")
+
+	return cache.dbInsertBucket(bucket)
+}
+
+// DeleteBucket will attempt to delete an existing bucket
+func (cache *SiaCacheLayer) DeleteBucket(bucket string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.DeleteBucket")
+
+	// Do NOT delete if objects exist in bucket.
+	objects, serr := cache.ListObjects(bucket)
+	if serr != nil {
+		return serr
+	}
+	if (len(objects) > 0) {
+		return siaErrorBucketNotEmpty
+	}
+
+	return cache.dbDeleteBucket(bucket)
+}
+
+// ListBuckets will return a list of all existing buckets
+func (cache *SiaCacheLayer) ListBuckets() (buckets []SiaBucketInfo, e *SiaServiceError) {
+	cache.debugmsg("SiaCacheLayer.ListBuckets")
+
+	return cache.dbListBuckets()
+}
+
+// DeleteObject will attempt to delete the object from Sia
+func (cache *SiaCacheLayer) DeleteObject(bucket string, objectName string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.DeleteObject")
+
+	err := cache.dbUpdateObjectDeletedStatus(bucket, objectName, 1)
+	if err != nil {
+		return err
+	}
+
+	// Tell Sia daemon to delete the object
+	var siaObj = cache.getSiaObjectName(bucket, objectName)
+
+	derr := post(cache.SiadAddress, "/renter/delete/"+siaObj, "")
+	if derr != nil {
+		return &SiaServiceError{Code: "SiaErrorDaemon", Message: derr.Error()}
+	}
+
+	return cache.dbDeleteObject(bucket, objectName)
+}
+
+// PutObject will attempt to put an object on Sia
+func (cache *SiaCacheLayer) PutObject(bucket string, objectName string, size int64, purgeAfter int64, srcFile string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.PutObject")
+
+	// Before inserting to DB, there is a very rare chance that the object already exists in DB
+	// from a failed upload and Minio crashed or was killed before DB updated to reflect. So just in case
+	// we will check if the object exists and has a not uploaded status. If so, we will delete that
+	// record and then continue as normal.
+	objInfo, e := cache.GetObjectInfo(bucket, objectName)
+	if e == nil {
+		// Object does exist. If uploaded, return error. If not uploaded, delete it and continue.
+		if objInfo.Uploaded.Unix() > 0 {
+			return siaErrorObjectAlreadyExists
+		}
+		e = cache.dbDeleteObject(bucket, objectName)
+		if e != nil {
+			return e
+		}
+	}
+
+	err := cache.dbInsertObject(bucket, objectName, size, time.Now().Unix(), 0, purgeAfter, srcFile, 1)
```

Comment:
Does this insert the entire object to `boltDB` ? if yes have you tested with large files? - Would it be possible to use boltDB just for namespace and use regular filesystem for staging? it would be faster..

Reply:
I see that its just namespace..

Reply:
Correct. DB is only used to store small metadata for buckets and objects.